pkg/api/types.go

```diff
@@ -1776,6 +1774,110 @@ type SecretList struct {
 	Items []Secret `json:"items"`
 }
 
+// SecurityContext holds security configuration that will be applied to a container.  If a security context
+// is set on the container spec then it must comply with any constraints defined in the SecurityConstraints context
+// it is running in.  If a security context is not supplied and the pod is running under a SecurityConstraints context
+// then a default SecurityContext may be applied.
+type SecurityContext struct {
+	// Capabilities are the capabilities to add/drop when running the container
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
+
+	// Run the container in privileged mode
+	Privileged bool `json:"privileged,omitempty"`
+
+	// SELinuxOptions are the labels to be applied to the container
+	// and volumes
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
+
+	// RunAsUser is the UID to run the entrypoint of the container process.  Corresponding option is --user or -u
+	RunAsUser int64 `json:"runAsUser,omitempty"`
+}
+
+// SELinuxOptions are the labels to be applied to the container
+type SELinuxOptions struct {
+	// User --security-opt="label:user:USER"
+	User string `json:"user,omitempty"`
+
+	// Role --security-opt="label:role:ROLE"
+	Role string `json:"role,omitempty"`
+
+	// Type --security-opt="label:type:TYPE"
+	Type string `json:"type,omitempty"`
+
+	// Level --security-opt="label:level:LEVEL"
+	Level string `json:"level,omitempty"`
+
+	// Disabled --security-opt="label:disable"
+	Disabled bool `json:"disabled,omitempty"`
+}
+
+// SecurityConstraints provides the constraints that at security context provider will
+// ensure that applied SecurityContext requests follow.  When a setting is provided in the SecurityConstraints
+// that conflicts with an actual request it will be implementation specific whether that request is
+// ignored and the container is still run (without the requested constraint) or if the container will be failed
+type SecurityConstraints struct {
+	// EnforcementPolicy will drive behavior for how the constraints are enforce
+	EnforcementPolicy SecurityConstraintPolicy `json:"enforcementPolicy,omitempty"`
+
+	// AllowPrivileged indicates whether this context allows privileged mode containers
+	AllowPrivileged bool `json:"allowPrivileged,omitempty"`
+
+	// SELinux provides the security constraint options for selinux
+	SELinux *SELinuxSecurityConstraints `json:"seLinux,omitempty"`
+
+	// AllowCapabilities dictates if a container can request to add or drop capabilites
+	AllowCapabilities bool `json:"allowCapabilities,omitempty"`
+
+	// AllowCapabilities dictates if a container can request to run the entry point process as a specific user
+	AllowRunAsUser bool `json:"allowRunAsUser,omitempty"`
```

Comment:
This is interesting, but there's probably a number of options here like:

* allow Run as User to be set to arbitrary
* allow run as user in this range
* allow run as non-root user
* allow run as with the image user

That means this probably needs to be a struct.